rfcs/0145-doc-strings.md

```diff
@@ -0,0 +1,519 @@
+    - (+) Further Arguments for CommonMark, in general, can be found in RFC72
+- (+) Allows copy-paste from and to markdown files. We allow easy refactoring if documentation arises and needs to be split into separate files.
+- (-) Strictly binding doc comments content to commonMark might restrict users.
+    - (+) Users/Tools can still use regular comments or develop alternative solutions.
+- (-) CommonMark does not specify the current rich features from nixdoc, such as predefined sections and structures that could be used as a source for automated toolings. Such as types for type-checking or examples to run automated tests
+    - (+) Future tools can still build their conventions on top of this RFC. They might not directly specify them in an RFC but be a tool developers choose for a specific codebase. However, we have yet to get those tools. So, it is good when rich features remain unspecified.
+  
+</details>
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+## Changes the existing comments inside the code base
+
+This could be (partially) automated. (see our [codemod](https://github.com/nix-community/docnix/tree/3c0531cb5b4c9f3e9069b73d19e6c4be8508d905/codemod) )
+
+Also, the migration could be performed piecemeal, starting perhaps with Nixpkgs `lib`.
+
+See future work.
+
+## Breaking the Nixpkgs manual
+
+This is a breaking change to the current Nixpkgs manual tooling Nixpkgs library function documentation.
+
+Please take a look at future work.
+
+# Alternatives
+[alternatives]: #alternatives
+
+## All considered outer formats
+
+| Property / Approach | `##` | `/** */` | `Javadoc` | `/*\|` or `/*^`  | All comments are doc-comments |
+|---|---|---|---|---|---|
+| Inspired by | Rust | Current Nixpkgs `lib` | C++/Java/JavaScript | Haskell Haddock | Current Nixpkgs.lib |
+| Changes the existing code by | Much | Less | Even More | Less | ? |
+| Needs Termination | No | Yes | Yes | Yes | ? |
+| Indentation | Clear | like Nix's multiline strings, thus **Intuitive** | Clear | ? | ? |
+| Needs vertical space  | No | Yes | Yes | Yes | ? |
+| Visual distinction from comments | High | Low | Medium | Medium | No distinction at all |
+| Needs Autocompletion (Language Server) to continue the next line. | Yes | No | Yes | No | ? |
+| Punctuation Variations / Amount of different special characters | 1 (Less) | 2 (Medium) | 2 (Medium) | 3 (More) | None |
+| Markdown compatibility (also depends on indentation clarity) | **Visual conflicts with headings `# Title`** | Good | Medium | Good | Only with multiline comments |
+| breaks when interrupted with newlines | Yes | No | ? | No | |
+| Simplicity (Brainload) | Medium | Simple | Complex | More Complex | |
+
+### Refactoring note
+
+**Observing**: From a refactoring perspective, it might also be interesting to see how many conflicts the different formats would cause.
+
+Nixpkgs comments:
+
+- `##` ~4k usages
+- `#` ~20k usages
+- `/*` ~6k usages
+- `/**` 160 usages 
```

Comment:
If this was unclear, it is what we expect to be a non-doc-comment only ``/***/` would be a doc-comment although it is empty.

Reply:
> This could be solved by prescribing the doc comment parser to somehow reject these kind of /**/ comments somehow.

Isn't `/**/` just a normal block(multi-line) comment in this case? I think doc-comment should be `\/\*\*([^*]|\*+[^*/])*\*+\/`, that is, it should at least consume `/**` and then ends with `*/`, without overlapping `*`. IMHO, this is very easy to impl in lexers.

Reply:
Granted, I just wanted to mention it, as `/**/` accounts for a significant amount of the occurrences of `/**` in nixpkgs, so needs to be kept in mind for this list.

Reply:
Just counted the occurrences of `/**/` as comments:
35 all of them are used for "alignments" like this:

```nix
  foo =
    /**/ if condition1 then expr
    else if condition2 then alternative
    else throw "error";
```

Since we don't consider them doc-comments, they should be ignored anyways. Although i think it makes sense to keep an overview, that it might have contributed to our decision to use `/** */`